bindings/bind_py_representation_calculator.cc

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file   bind_py_representation_calculator.cc
+ *
+ * @author Felix Musil <felix.musil@epfl.ch>
+ *
+ * @date   30 Oct 2018
+ *
+ * @brief  File for binding the Representation calculator
+ *
+ * Copyright  2018  Felix Musil, COSMO (EPFL), LAMMM (EPFL)
+ *
+ * Rascal is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * Rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Emacs; see the file COPYING. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#include "bind_py_representation_calculator.hh"
+
+namespace rascal {
+
+  template <class Calculator>
+  using PyCalculator = py::class_<Calculator, CalculatorBase>;
+
+  template <typename Calculator>
+  decltype(auto) add_representation_calculator(py::module & mod, py::module &) {
+    std::string representation_name =
+        internal::GetBindingTypeName<Calculator>();
+
+    py::class_<Calculator, CalculatorBase> representation(
+        mod, representation_name.c_str());
+    // use custom constructor to pass json formated string as initializer
+    // an alternative would be to convert python dict to json internally
+    // but needs some work on in the pybind machinery
+    representation.def(py::init([](std::string & hyper_str) {
+      // convert to json
+      json hypers = json::parse(hyper_str);
+      return std::make_unique<Calculator>(hypers);
+    }));
+
+    return representation;
+  }
+
+  template <class Calculator, typename Manager,
+            template <class> class... Adaptor>
+  void bind_compute_function(PyCalculator<Calculator> & representation) {
+    using TypeHolder_t = StructureManagerTypeHolder<Manager, Adaptor...>;
+    using Manager_t = typename TypeHolder_t::type;
+    representation.def(
+        "compute", &Calculator::template compute<std::shared_ptr<Manager_t>>,
+        py::call_guard<py::gil_scoped_release>());
+
+    representation.def(
+        "compute",
+        &Calculator::template compute<ManagerCollection<Manager, Adaptor...>>,
+        py::call_guard<py::gil_scoped_release>());
+  }
+
+  namespace detail {
```

Comment:
Haven't we been using `internal` as the namespace for implementation details like this? Maybe `py_internal` if you want to separate the C++ and py-bind internals?

Reply:
does it make a difference ? both namespace are used in the cpp.